HW2/P2/parallel_vector.pyx

```diff
@@ -70,23 +70,40 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+  cdef:
+    int idx, r, minlock, maxlock
+    omp_lock_t *locks = get_N_locks(counts.shape[0])
+
+  ##########
+  # Your code here
+  # Use parallel.prange() and a lock for each element of counts to parallelize
+  # data movement.  Be sure to avoid deadlock, and double-locking.
+  ##########
+  with nogil:
+    for r in range(repeat):
+      # loop in parallel with 4 threads
+      for idx in prange(src.shape[0], num_threads=4):
+      	# need to make sure we lock in order! always lock the smaller idx first
+        minlock = min(src[idx], dest[idx])
+        maxlock = max(src[idx], dest[idx])
+
+        # if the lock is the same, great, don't need to lock both locks!
+        if minlock == maxlock:
+          omp_set_lock(&locks[minlock])
+          if counts[src[idx]] > 0:
+            counts[dest[idx]] += 1
+            counts[src[idx]] -= 1
+          omp_unset_lock(&locks[minlock])
+        else:
+          omp_set_lock(&locks[minlock])
+          omp_set_lock(&locks[maxlock])
+          if counts[src[idx]] > 0:
+            counts[dest[idx]] += 1
+            counts[src[idx]] -= 1
+          omp_unset_lock(&locks[minlock])
+          omp_unset_lock(&locks[maxlock])
+
+  free_N_locks(counts.shape[0], locks)
 
```

Comment:
This is great that you considered many cases when posing the lock. I understand that if src and dest are the same we only need to lock once, but do we really have to always lock the small one first? I don't see any necessities here since it's just data movement that doesn't involve any comparisons